src/Processors/Formats/Impl/Parquet/ParquetDataValuesReader.cpp

```diff
@@ -0,0 +1,570 @@
+            }
+        }
+        else
+        {
+            repeated_visitor(cur_count, cur_value);
+        }
+        cur_group_cursor += cur_count;
+        num_values -= cur_count;
+    }
+}
+
+template <typename IndividualVisitor, typename RepeatedVisitor>
+void RleValuesReader::visitNullableValues(
+    size_t cursor,
+    UInt32 num_values,
+    Int32 max_def_level,
+    LazyNullMap & null_map,
+    IndividualVisitor && individual_visitor,
+    RepeatedVisitor && repeated_visitor)
+{
+    while (num_values)
+    {
+        nextGroupIfNecessary();
+        auto cur_count = std::min(num_values, curGroupLeft());
+
+        if (cur_group_is_packed)
+        {
+            for (auto i = cur_group_cursor; i < cur_group_cursor + cur_count; i++)
+            {
+                if (cur_packed_bit_values[i] == max_def_level)
+                {
+                    individual_visitor(cursor);
+                }
+                else
+                {
+                    null_map.setNull(cursor);
+                }
+                cursor++;
+            }
+        }
+        else
+        {
+            if (cur_value == max_def_level)
+            {
+                repeated_visitor(cursor, cur_count);
+            }
+            else
+            {
+                null_map.setNull(cursor, cur_count);
+            }
+            cursor += cur_count;
+        }
+        cur_group_cursor += cur_count;
+        num_values -= cur_count;
+    }
+}
+
+template <typename IndividualNullVisitor, typename SteppedValidVisitor, typename RepeatedVisitor>
+void RleValuesReader::visitNullableBySteps(
+    size_t cursor,
+    UInt32 num_values,
+    Int32 max_def_level,
+    IndividualNullVisitor && individual_null_visitor,
+    SteppedValidVisitor && stepped_valid_visitor,
+    RepeatedVisitor && repeated_visitor)
+{
+    // refer to:
+    // RleDecoder::GetBatch in rle_encoding.h and TypedColumnReaderImpl::ReadBatchSpaced in column_reader.cc
+    // VectorizedRleValuesReader::readBatchInternal in Spark
+    while (num_values > 0)
+    {
+        nextGroupIfNecessary();
+        auto cur_count = std::min(num_values, curGroupLeft());
+
+        if (cur_group_is_packed)
+        {
+            valid_index_steps.resize(cur_count + 1);
+            valid_index_steps[0] = 0;
+            auto step_idx = 0;
+            auto null_map_cursor = cursor;
+
+            for (auto i = cur_group_cursor; i < cur_group_cursor + cur_count; i++)
+            {
+                if (cur_packed_bit_values[i] == max_def_level)
+                {
+                    valid_index_steps[++step_idx] = 1;
+                }
+                else
+                {
+                    individual_null_visitor(null_map_cursor);
+                    if (unlikely(valid_index_steps[step_idx] == UINT8_MAX))
+                    {
+                        throw Exception(ErrorCodes::PARQUET_EXCEPTION, "unsupported packed values number");
```

Comment:
Why is this unexpected? Because it would be encoded as repeated run instead of packed? But the definition levels may be different, even if they're all less than max_def_level.

E.g. for this Array(Nullable): `[null], [], [null], [], [null], [], ...` the definition levels are (I'm not certain): `1, 0, 1, 0, 1, 0, ...`, and `max_def_level = 2`, so it'll hit this exception.

Currently it's ok because arrays are not supported here, but please at least add a TODO to change this before adding arrays.

Reply:
The element in `valid_index_steps` describes how many elements there are from i-th valid element (include) to (i+1)-th valid element (exclude), which can be found in comments of `visitNullableBySteps` in ParquetDataValuesReader.h. So, it's the max consecutive number that is stored in `valid_index_steps`, which will not be bigger than UNIT8_MAX.

The check here is used as a protection for bad data, and the performance impaction is ignorable also.

Reply:
Yes, obviously `valid_index_steps` can't contain numbers bigger than 255. But that's a limitation that you imposed, not a limitation in the parquet format, right? If a valid parquet file can contain more than 255 consecutive non-max def levels in a packed run, this code will incorrectly fail to parse it.

https://parquet.apache.org/docs/file-format/data-pages/encodings/ says "bit-packed-run-len and rle-run-len must be in the range [1, 2^31 - 1]" and doesn't mention any additional smaller limit for packed runs. So I guess the runs can be longer than 255 even in practice.

You can also just try to break it with: `select multiIf(number=0, [1], number%2=0, [], [null::Nullable(Int8)]) as a from numbers(100000) into outfile 't.parquet'` (but if this doesn't break it, maybe something else will).

Seems easy to fix too: when hitting 255, call the callback, and clear valid_index_steps.

Reply:
Indeed, all current implementations will not write more than 8 consecutive values in bit packed format, this can be verified by[ C++ source code](https://github.com/ClickHouse/arrow/blob/8f36d71d18587f1f315ec832f424183cb6519cbb/cpp/src/arrow/util/rle_encoding.h#L674), and [Java source code](https://github.com/apache/parquet-mr/blob/77fa24eb25442af5ca9acf591d1b37c02db2b037/parquet-column/src/main/java/org/apache/parquet/column/values/rle/RunLengthBitPackingHybridEncoder.java#L146).

So, current 255 limitation is enough. Further more, it's more reasonable to use RLE format for more than 8 consecutive values instead of bit packed format. So, we should not implement any protection code any more just according to the definition. And if we implemented the fallback logic, we can not even generate a test parquet file by any valid Parquet writer.

Reply:
Thanks for pointing to the code.

If I'm reading it correctly, it can write up to 63*8 (>255!) values in one packed run: https://github.com/ClickHouse/arrow/blob/8f36d71d18587f1f315ec832f424183cb6519cbb/cpp/src/arrow/util/rle_encoding.h#L769-L770 . buffered_values_ holds at most 8 values at a time, and FlushBufferedValues/FlushLiteralRun writes them out, BUT it doesn't necessarily start a new packed run every time, it can append to an existing packed run, up to 63*8 values.

> it's more reasonable to use RLE format for more than 8 consecutive values instead of bit packed format

Only if the values are equal. If they're different (but all less than max_def_level) then they can't be a RLE run.

Reply:
May be my comments in source code are not clear enough, I changed the comments of `visitNullableBySteps`, and `valid_index_steps` records the **gap size** between two valid elements.

So, the total elements number in a packed group do not affect the value in `visitNullableBySteps`, while only the max consecutive values do. So the current implementation is OK.

Reply:
What I'm saying is that it's possible for a packed run to contain 63*8 consecutive invalid elements. I'm running out of ways to say it more clearly :(

Reply:
> (But this case is not reachable right now because arrays and nested nullables are not supported, so max_def_level is at most 1. Maybe that's what you meant?)

OK, finally, may be this is the point.
Currently, nested type is not supported, so all invalid elements are corresponding to same definition level. So consecutive invalid elements imply consecutive definition level values, which should not be more than 8, **currently**.

By the way, the logic about `valid_index_steps` must be refactored if the nested type supported. Because for consecutive invalid elements with different definition values, we can not visit them with a loop of same logic, instead, we have to update the corresponding nest level object according to different *invalid* definition level value.

Reply:
I hacked the code to allow reading arrays (just reading the inner primitive column, without turning it into array), and it does run into this UINT8_MAX limit:

```
--- a/src/Processors/Formats/Impl/Parquet/ParquetRecordReader.cpp
+++ b/src/Processors/Formats/Impl/Parquet/ParquetRecordReader.cpp
@@ -329,8 +329,8 @@ ParquetRecordReader::ParquetRecordReader(
             throw Exception(ErrorCodes::PARQUET_EXCEPTION, "no column with '{}' in parquet file", col_with_name.name);
 
         auto node = it->second;
-        if (!node->is_primitive())
-            throw Exception(ErrorCodes::NOT_IMPLEMENTED, "arrays and maps are not implemented in native parquet reader");
+        while (!node->is_primitive())
+            node = dynamic_cast<parquet::schema::GroupNode*>(node.get())->field(0);
 
         auto idx = file_reader->metadata()->schema()->ColumnIndex(*node);
         chassert(idx >= 0);
```

```
:) insert into function file('y.parquet') select multiIf(number=0, [1], number%2=0, [], [null])::Array(Nullable(Int64)) as a from numbers(1000) settings output_format_parquet_use_custom_encoder=0

INSERT INTO FUNCTION file('y.parquet')
SETTINGS output_format_parquet_use_custom_encoder = 0
SELECT CAST(multiIf(number = 0, [1], (number % 2) = 0, [], [NULL]), 'Array(Nullable(Int64))') AS a
FROM numbers(1000)
SETTINGS output_format_parquet_use_custom_encoder = 0

Query id: f2a65fda-a47a-404e-b7fb-dd1bd0aaa4bb

Ok.

0 rows in set. Elapsed: 0.057 sec. Processed 1.00 thousand rows, 8.00 KB (17.46 thousand rows/s., 139.68 KB/s.)
Peak memory usage: 12.20 KiB.

:) select * from file('y.parquet') limit 10 settings input_format_parquet_use_native_reader=1

SELECT *
FROM file('y.parquet')
LIMIT 10
SETTINGS input_format_parquet_use_native_reader = 1

Query id: 374a7a84-b59e-4965-af37-a68cea9e6d37


Elapsed: 0.021 sec. 

Received exception from server (version 24.5.1):
Code: 722. DB::Exception: Received from localhost:9000. DB::Exception: unsupported packed values number: (in file/uri /home/ubuntu/cluster/n1/store/user_files/y.parquet): While executing ParquetBlockInputFormat: While executing File. ()
```

---

Now that I think about it, this code will probably have to be rewritten when implementing arrays anyway, so maybe indeed there's no point in fixing it now. But at least please add a TODO to make sure the array implementation handles this case correctly!

Reply:
OK, deal :).
I will add TODO